File: Chapter-7/queue_imp_using_linkedList.cpp
#include<bits/stdc++.h>

using namespace std;

struct Node{
    int value;
    struct Node *link;
};

struct Node *front = 0;
struct Node *rear = 0;

void EnQueue(){
    int data;
    cout<<" enter the data: ";
    cin>>data;
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->value = data;
    newNode->link =NULL;
    if(front == 0 && rear ==0){
        rear = newNode;
        front = newNode;
    }
    else{
        rear->link = newNode;
        rear = newNode;
    }
}

void DeQueue(){
    struct Node *item;
    if(front == 0 && rear == 0){
        cout<<" Underflow "<<endl;
    }
    else if(rear == front){
        cout<<"Dequeue element is: "<<front->value<<endl;
        rear = front =0;
    }
    else{
        item = front;
        cout<<"Dequeue element is: "<<front->value<<endl;
        front = front->link;
        free(item);
    }
}

void Peek_Queue(){
    if(front == 0 && rear == 0){
        cout<<" Underflow "<<endl;
    }
    else{
        cout<<"Peek element is: "<<front->value<<endl;
    }
}

void Display_Queue(){
    struct Node *temp = front;
    if(front == 0 && rear == 0){
        cout<<" Underflow "<<endl;
    }
    while(temp!=NULL){
        cout<<temp->value<<" ";
        temp = temp->link;
    }
}


int main(){
    int task;
    do{
        cout<<"1. Enqueue in queue"<<endl;
        cout<<"2. dequeue in a queue"<<endl;
        cout<<"3. peak in queue"<<endl;
        cout<<"4. Display The queue"<<endl;
        cout<<"0. For Exit"<<endl;
        cout<<"Enter an option : ";
        cin>>task;
        switch(task){
        case 0:
            break;
        case 1:
            EnQueue();
            cout<<endl;
            break;

        case 2:
            DeQueue();
            cout<<endl;
            break;

        case 3:
            Peek_Queue();
            cout<<endl;
            break;

        case 4:
            Display_Queue();
            cout<<endl;
            break;

        default:
            cout<<"please enter the correct option"<<endl;
            cout<<endl;
            break;
        }
    }while(task!=0);

    return 0;
}

File: Chapter-7/queue_imp_using_linkedList.py
class Node1:
    def __init__(self, item):
        self.item = item
        self.link = None


class Que:
    def __init__(self):
        self.front = 0
        self.rear = 0

    def EnQueue(self, key):
        newNode = Node1(key)
        newNode.link = None
        if self.front == 0 and self.rear == 0:
            self.front = self.rear = newNode
        else:
            self.rear.link = newNode
            self.rear = newNode

    def DeQueue(self):
        if self.front == 0 and self.rear == 0:
            print(" Underflow!! ")
        elif self.front == self.rear:
            print(f"Dequeue element : {self.front.item}")
            self.front = self.rear = 0

        else:
            print(f"Dequeue element : {self.front.item}")
            self.front = self.front.link

    def Peek_Que(self):
        if self.front == 0 and self.rear == 0:
            print(" Underflow!! ")
        else:
            print(f"Peek element is: {self.front.item}")

    def Display_Que(self):
        temp = self.front
        if self.front == 0 and self.rear == 0:
            print(" Underflow!! ")
        else:
            print("ElementS in the queue are : ", end=" ")
            while temp is not None:
                print(temp.item, end=" ")
                temp = temp.link
            print()


Q = Que()
Q.EnQueue(1)
Q.EnQueue(90)
Q.EnQueue(87)
Q.EnQueue(23)
Q.EnQueue(56)
Q.EnQueue(32)
Q.EnQueue(34)
Q.Display_Que()
Q.DeQueue()
Q.Display_Que()
Q.Peek_Que()
